android/src/main/java/com/twiliorn/library/CustomVideoProcessor.java

```diff
@@ -0,0 +1,100 @@
+package com.twiliorn.library;
+
+import android.content.Context;
+import android.util.Log;
+
+import androidx.annotation.Nullable;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import tvi.webrtc.VideoFrame;
+import tvi.webrtc.VideoProcessor;
+import tvi.webrtc.VideoSink;
+
+/**
+ * This custom video processor allows us to capture frames before they are adapted & rendered to the screen.
+ * https://github.com/twilio/video-quickstart-android/blob/master/exampleAdvancedCameraCapturer/src/main/java/com/twilio/video/examples/advancedcameracapturer/Photographer.kt
+ */
+public class CustomVideoProcessor implements VideoProcessor {
+    private AtomicBoolean canCapture = new AtomicBoolean(true);
+    private AtomicBoolean captureThisFrame = new AtomicBoolean(false);
+    private PatchedVideoView videoView;
+    private Context context;
+    // choosing to use single thread since limited i/o resources may be a bottleneck anyways
+    private ExecutorService executorService = Executors.newSingleThreadExecutor();
+
+    public CustomVideoProcessor() {}
+
+    public void captureFrame() {
+        if (canCapture.get()) {
+            Log.d(TwilioPackage.TAG, "Setting captureThisFrame flag to true.");
+            captureThisFrame.set(true);
+        }
+        else {
+            Log.e(TwilioPackage.TAG, "Cannot capture frame. Capturer is not started.");
+        }
+    }
+
+    public void setVideoView(PatchedVideoView videoView) {
+        this.videoView = videoView;
+    }
+
+    public void setContext(Context context) {
+        this.context = context;
+    }
+
+    @Override
+    public void onFrameCaptured(VideoFrame frame, FrameAdaptationParameters parameters) {
+        if (frame == null) {
+            Log.w(TwilioPackage.TAG, "Cannot process null frame.");
+            return;
+        }
+
+        frame.retain();
+
+        if (captureThisFrame.get()) {
```

Comment:
Use `if (captureThisFrame.compareAndSet(true, false))` (and remove the separate `captureThisFrame.set(false)` below) so that the toggle of the value is atomic.

Reply:
Ok great advice 👍